src/data/real/basic.lean

```diff
@@ -13,17 +13,38 @@ import algebra.star.basic
 
 /-- The type `ℝ` of real numbers constructed as equivalence classes of Cauchy sequences of rational
 numbers. -/
-def real := @cau_seq.completion.Cauchy ℚ _ _ _ abs _
+structure real := of_cauchy ::
+(cauchy : @cau_seq.completion.Cauchy ℚ _ _ _ abs _)
 notation `ℝ` := real
 
 namespace real
 open cau_seq cau_seq.completion
 
 variables {x y : ℝ}
 
-def comm_ring_aux : comm_ring ℝ := Cauchy.comm_ring
-
-instance : comm_ring ℝ := { ..comm_ring_aux }
+lemma ext_cauchy_iff : ∀ {x y : real}, x = y ↔ x.cauchy = y.cauchy
+| ⟨a⟩ ⟨b⟩ := by split; cc
+
+lemma ext_cauchy {x y : real} : x.cauchy = y.cauchy → x = y :=
+ext_cauchy_iff.2
+
+instance : comm_ring ℝ :=
+{ zero := ⟨0⟩,
+  one := ⟨1⟩,
+  add := λ ⟨a⟩ ⟨b⟩, ⟨a + b⟩,
+  neg := λ ⟨a⟩, ⟨-a⟩,
+  mul := λ ⟨a⟩ ⟨b⟩, ⟨a * b⟩,
+  zero_add := λ ⟨a⟩, ext_cauchy (zero_add _),
+  add_zero := λ ⟨a⟩, ext_cauchy (add_zero a),
+  add_assoc := λ ⟨a⟩ ⟨b⟩ ⟨c⟩, ext_cauchy (add_assoc _ _ _),
+  add_comm := λ ⟨a⟩ ⟨b⟩, ext_cauchy (add_comm _ _),
+  add_left_neg := λ ⟨a⟩, ext_cauchy (add_left_neg _),
+  one_mul := λ ⟨a⟩, ext_cauchy (one_mul _),
+  mul_one := λ ⟨a⟩, ext_cauchy (mul_one a),
+  mul_assoc := λ ⟨a⟩ ⟨b⟩ ⟨c⟩, ext_cauchy (mul_assoc _ _ _),
+  mul_comm := λ ⟨a⟩ ⟨b⟩, ext_cauchy (mul_comm _ _),
+  left_distrib := λ ⟨a⟩ ⟨b⟩ ⟨c⟩, ext_cauchy (left_distrib _ _ _),
+  right_distrib := λ ⟨a⟩ ⟨b⟩ ⟨c⟩, ext_cauchy (right_distrib _ _ _) }
```

Comment:
It would be cool if we had a derive handler for `ring` etc on structures, which just defined operations field-wise